feat: Expose sub-query DocumentNode to RemoteGraphQLDataSource#willSendRequest hook

theJC opened the pull request:
Fixes #1352

Initial proposed change to provide access to the AST so that customer code processing a subgraph query no longer has to waste cycles re-parsing it back into a document when needed.

At the moment there are failing tests that will need to be updated to include the additional operationDocumentNode in the returned but I did not want to undergo that effort until we were sure that this was an agreeable go-forward direction to support this capability.

CC: @lennyburdette - here is the MR per your recommendation

theJC commented:
Adjusted.

Whats your thoughts on the best way to approach updating testing so it passes?  Are we best served by adding the representation of the documentNode into all the query plan results in build-query-plan.feature?

https://app.circleci.com/pipelines/github/apollographql/federation/8610/workflows/bf18dcef-c4cf-45e4-8ec0-b592a80f8e24/jobs/52912?invite=true#step-107-8795

pcmanus commented:
@theJC As the patch stand,  I don't know of other option than adding the representations all over in `build-query-plan.feature` (but to be honest, my knowledge of cucumber, used by those test, is weak). I do get it's annoying, both now and frankly moving forward since any future change will now have to be done in 2 places; and the value is unclear.

That said, following on my previous suggestion, one option might be to do "add an option to disable this" now. This would mean adding some new `options` argument to the `QueryPlanner` ctor, which currently has no options (but we'll likely have need for other options soon enough, so that's fine imo). If populating the new field can be disabled, then it could be used that to disable it for all the cucumber tests. It'd make sense in that case to have a dedicated non-cucumber unit-test that target this new addition, but at least the existing cucumber tests won't have to be updated.

Personally, I kind of prefer that later option because as said before, I suspect we'll have some use case at some point where generated this somewhat redundant info is not desirable. But I won't object to just updating all the cucumber tests either.

theJC commented:
@pcmanus I believe I've addressed your concerns, let me know if there is anything else needed to make this PR mergeable.

Thanks again for your input,
Jon

pcmanus reviewed (approved):
Sorry the delay. Patch looks great, thanks, though I'm afraid it needs rebasing (one of the reason is that the `buildPlan.ts` file has been reorganised a bit, which confuses the diff algorithm). It'd be awesome if you could:
1. rebase it.
2. add a quick line in `query-planner/CHANGELOG.md`, and copy-paste it also in `gateway-js/CHANGELOG.md`.

and then we can ship it.

theJC commented:
Should be all set @pcmanus

pcmanus commented:
Perfect, thanks a lot of the contribution!

pcmanus commented:
Hi @theJC. I wanted to let you know that we're planning to switch this option to `false` by default in the upcoming 2.2 release (see #2227/#2230), so you would have to manually opt-in as you upgrade if you still rely on this. Sorry for the inconvenience.
